memgpt/local_llm/chat_completion_proxy.py

```diff
@@ -88,11 +81,29 @@ def get_chat_completion(
     # User provided a correct prompt formatter
     else:
         llm_wrapper = available_wrappers[wrapper]
-        # TODO move this to a flag
-        if "_grammar" in wrapper:
-            setattr(llm_wrapper, "assistant_prefix_extra_first_message", "")
-            setattr(llm_wrapper, "assistant_prefix_extra", "")
-            grammar, documentation = generate_grammar_and_documentation(functions_python)
+
+    # If the wrapper uses grammar, generate the grammar using the grammar generating function
+    # TODO move this to a flag
+    if "_grammar" in wrapper:
+        # When using grammars, we don't want to do any extras output tricks like appending a response prefix
+        setattr(llm_wrapper, "assistant_prefix_extra_first_message", "")
+        setattr(llm_wrapper, "assistant_prefix_extra", "")
+
+        # TODO find a better way to do this than string matching (eg an attribute)
+        if "noforce" in wrapper:
+            # "noforce" means that the prompt formatter expects inner thoughts as a top-level parameter
+            # this is closer to the OpenAI style since it allows for messages w/o any function calls

```

Comment:
@Maximilian-Winter is this true with the way the grammar generator is written? ie, will it allow output like this?

```json
{
  "inner_thoughts": "Well, nothing much to do, so I won't call any functions"
}
```

That's technically possible with the `chatml-noforce` wrapper (when not using grammars).

Reply:
Right now it is not possible, but that would be an easy change to made.

Reply:
Added the functionality.